File: frontend/types/paper.ts
export type PaperStatus = 'pending' | 'processing' | 'completed' | 'failed';

export interface PaperSummary {
  id: string;
  workspace_id: string;
  title: string;
  filename: string;
  status: PaperStatus;
  created_at: string;
  updated_at: string;
}

export interface StructuredOutput {
  summary: string;
  sections: { title: string; summary: string }[];
  key_insights: string[];
  novel_contributions: string[];
  limitations: string[];
  math_explanations: { concept: string; formula: string; variable_notes: string[]; explanation: string; importance: string }[];
  implementation_steps: { step: string; detail: string }[];
  sandbox_tasks: { title: string; objective: string }[];
  starter_code: string;
  qa_ready: boolean;
  fallback_mode?: boolean;
}

export interface QAResponse {
  status: 'stated' | 'inferred' | 'not_stated' | 'insufficient_evidence';
  question_type: 'direct_fact' | 'synthesis' | 'inference' | 'missing_info';
  answer: string;
  evidence: { source: string; passage: string }[];
  confidence: number;
  citations: string[];
}

export interface PaperDetail {
  paper: PaperSummary;
  output: StructuredOutput | null;
  sandbox: {
    current_code: string;
    starter_code: string;
    last_run_output: { stdout: string; stderr: string; success: boolean };
  } | null;
}

File: backend/app/schemas/paper.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class PaperSummary(BaseModel):
    id: str
    workspace_id: str
    title: str
    filename: str
    status: Literal["pending", "processing", "completed", "failed"]
    created_at: datetime
    updated_at: datetime


class UploadPaperResponse(BaseModel):
    paper: PaperSummary
    job: dict
    already_exists: bool = False
    message: str | None = None


class ReprocessPaperResponse(BaseModel):
    paper: PaperSummary
    job: dict
    message: str


class JobStatusResponse(BaseModel):
    id: str
    paper_id: str
    status: Literal["pending", "processing", "completed", "failed"]
    attempt_count: int
    error: str | None = None
    updated_at: datetime


class SectionModel(BaseModel):
    title: str
    summary: str


class MathExplanationModel(BaseModel):
    concept: str
    formula: str = ""
    variable_notes: list[str] = Field(default_factory=list)
    explanation: str
    importance: str = ""
    source_type: str = "text"
    source_context: str = ""


class ImplementationStepModel(BaseModel):
    step: str
    detail: str


class SandboxTaskModel(BaseModel):
    title: str
    objective: str


class StructuredPaperOutput(BaseModel):
    summary: str
    sections: list[SectionModel] = Field(default_factory=list)
    key_insights: list[str] = Field(default_factory=list)
    novel_contributions: list[str] = Field(default_factory=list)
    limitations: list[str] = Field(default_factory=list)
    math_explanations: list[MathExplanationModel] = Field(default_factory=list)
    implementation_steps: list[ImplementationStepModel] = Field(default_factory=list)
    sandbox_tasks: list[SandboxTaskModel] = Field(default_factory=list)
    starter_code: str
    qa_ready: bool
    fallback_mode: bool = False


class PaperDetailResponse(BaseModel):
    paper: PaperSummary
    output: StructuredPaperOutput | None = None
    sandbox: dict | None = None
